### ISSUE TITLE
PVF validation host: pull not push PVFs

### ISSUE THREAD

**Issue by pepyakin** [paritytech/polkadot#2866]
Currently candidate validation pushes PVF preimages to the candidate validation host. This is non-ideal, because PVF validation host caches compiled PVFs and does not need the original code blobs.

To fix this we should update PVF validation host so that it accepts not only raw PVFs but also PVF code hashes. In case of a cache miss, the PVF validation host should request the validation code preimage from candidate-validation.

It seems we cannot completely get rid of the mode where PVFs are passed directly as preimages. For example, for disputes, the subsystem that initiates candidate validation will have the validation code at hand, however, there is no guarantee the code can be found on the chain. In this case, it is just more convenient to pass the validation code directly.

**Comment by burdges** [comment]
We do not yet move code into availability with this one, right?

> for disputes, the subsystem that initiates candidate validation will have the validation code at hand, however, there is no guarantee the code can be found on the chain. In this case, it is just more convenient to pass the validation code directly.

This is true, but if we do code via availability then we could guarantee code can be found.  As disputes retrigger availability for the candidate, then could also retrigger availability for the code.  We expect both to already be availability so this usually resolves with a vote. 

I'm not sure about the best steps to take to reach this however.

**Comment by burdges** [comment]
Also @rphmeier we discussed whether approval checkers should reconstruct and build these special thin code blocks.  You convinced me approval checkers need not build the code, since if the code is invalid then every candidate becomes invalid, but we do still need approval checkers to check the erasure code reconstruction, or else someone can hide two different PVFs in the same erasure code data.

**Comment by pepyakin** [comment]
> We do not yet move code into availability with this one, right?

Yes. In fact, I see this issue as a rather tactical change rather than strategical.

**Comment by eskimor** [comment]
> It seems we cannot completely get rid of the mode where PVFs are passed directly as preimages. For example, for disputes, the subsystem that initiates candidate validation will have the validation code at hand, however, there is no guarantee the code can be found on the chain. In this case, it is just more convenient to pass the validation code directly.

I think we can and should. For disputes we always query the chain for the PVF, so we very much rely on the PVF being on chain - and it will be, otherwise it won't be a valid dispute. If we want to be able to dispute non included candidates in the future, the mechanism for doing that will be quite different and we can just make sure to have the PVF when the validation host requests them, after we passed the hash.

I would very much like to avoid the added complexity of supporting both push and pull mechanisms.

**Comment by pepyakin** [comment]
Agreed and good call.

I think this just got here because of lack of full grasp of the system at the moment of writing.
